Add alias to hidden in query options

MattAgn opened the pull request:
This PR aims to tackle [this issue](https://github.com/callstack/react-native-testing-library/issues/1194). Basically the idea is to add a new option `includeHidden` that will be an alias to the query option `hidden` but much clearer. We keep the old name `hidden` for compatibility with RTL

### Summary

- [x] Alias `hidden` query option to `hidden`
- [x] Alias `defaultHidden` config to `defaultIncludeHidden`
- [x] add new alias to docs

### Test plan

- [x] Check new `includeHidden` option is working for all queries
- [x] Check new `defaultIncludeHidden` option is working

### Remaining work

- [x] Choose a naming between `includeHidden` and `includeHiddenFromAccessibility`

Resolves #1194

MattAgn commented:
@mdjastrzebski I've got some doubts about my jsdocs for the `hidden` alias, is it clear?
I was also wondering how to deal with the fact that we have two options with the same name. Ideally, we'd want to have both options in the api but internally to handle only one. However, it seems much more simple to handle everything in the `findAll` (both for the global config and the queries). What are your thoughts on that?

mdjastrzebski commented on src/helpers/findAll.ts:
```diff
@@ -12,8 +13,11 @@ export function findAll(
 ) {
   const results = root.findAll(predicate);
+  const includeHiddenQueryOption = options?.includeHidden ?? options?.hidden;
+  const defaultIncludeHidden =
+    getConfig()?.defaultIncludeHidden ?? getConfig()?.defaultHidden;
 
-  const hidden = options?.hidden ?? getConfig().defaultHidden;
+  const hidden = includeHiddenQueryOption ?? defaultIncludeHidden;
```
```
const includeHidden = 
  options?.includeHidden ??
  options?.hidden ??
  getConfig()?.defaultIncludeHidden ??
  getConfig()?.defaultHidden
 }

```

MattAgn replied:
i wanted to avoid nested operators to maximise readibility, do you find it more readable your way?

mdjastrzebski replied:
Yes, I find it much more readable, as a single nullish coalescing operator. Having it as separate intermediate vars would be useful if it used a mix of two operators (like `&&` and `||`) with different precedence, etc. In case all vars use the same operator, I found putting it together removes the question why are there some intermediate variables :-)

MattAgn commented:
@mdjastrzebski I pushed a new commit with the other naming `includeHiddenFromAccessibility` so we can compare
Which do you prefer?

mdjastrzebski commented:
@MattAgn `includeHiddenElements` seems like a reasonable option. I would give it a go an see how the resulting code looks.

mdjastrzebski commented on src/helpers/findAll.ts:
```diff
@@ -12,8 +13,13 @@ export function findAll(
   options?: FindAllOptions
 ) {
   const results = root.findAll(predicate);
+  const includeHiddenElementsQueryOption =
```
The following should be easier to read and give the same result
```ts
const includeHiddenElements = 
  options?.includeHiddenElements ??
  options?.hidden ??
  getConfig()?.defaultIncludeHiddenElements ??
  getConfig()?.defaultHidden;
```

mdjastrzebski commented on typings/index.flow.js:
```diff
@@ -8,7 +8,10 @@ type QueryAllReturn = Array<ReactTestInstance> | [];
 
-type CommonQueryOptions = { hidden?: boolean };
+type CommonQueryOptions = {
+  hidden?: boolean,
+  includeHiddenElements?: boolean,
```
```suggestion
  includeHiddenElements?: boolean,
  
  /** Alias to `includeHiddenElements` for RTL compatibility */
  hidden?: boolean,

```

mdjastrzebski commented on typings/index.flow.js:
```diff
@@ -463,6 +466,8 @@ declare module '@testing-library/react-native' {
 
   declare interface Config {
     asyncUtilTimeout: number;
+    defaultHidden: boolean;
+    defaultIncludeHiddenElements: boolean;
```
```suggestion
    defaultIncludeHiddenElements: boolean;
    defaultHidden: boolean;
```

mdjastrzebski commented on website/docs/API.md:
```diff
@@ -791,9 +792,13 @@ function configure(options: Partial<Config>) {}
+
+Default [includeHiddenElements](Queries.md#includehidden-option) query option for all queries. This default option will be overridden by the one you specify directly when using your query.
+
 #### `defaultHidden` option
```
I would rather not give it a full header but rather put a mention in `defaultIncludeHiddenElements` entry that there is an alias for RTL compatibility.

mdjastrzebski commented on website/docs/Queries.md:
```diff
@@ -389,18 +390,22 @@ An element is considered to be hidden from accessibility based on [`isHiddenFrom
 
+### `hidden` option
+
+This is just an alias to the `includeHiddenElements` option. It only exists to match react-testing-library naming used in [byRole](https://testing-library.com/docs/queries/byrole/#hidden). Prefer the use of `includeHiddenElements` if possible as `includeHiddenElements: true` is clearer than `hidden: true`.
```
Similar to `configure` comment, I would rather not give it a full option header but only mention it under   `includeHiddenElements` option docs.

mdjastrzebski reviewed (commented):
I like the new name `includeHiddenElements` seems to strike a proper balance between being readable and concise. As they say "naming is hard" but I think you nailed it 👍🏻 

Regarding usage I would make sure we emphasis the readable variant, i.e. put it first in relevant tests, examples, docs, etc.

At this stage the goal for `hidden` alias is to help folks migrating from RTL or building dual RNTL/RTL wrappers. So it should work but should not be prominent at all not to confuse the main audience.

mdjastrzebski commented on src/queries/__tests__/a11yState.test.tsx:
```diff
@@ -239,9 +239,18 @@ test('byA11yState queries support hidden option', () => {
 
   expect(getByA11yState({ expanded: false })).toBeTruthy();
   expect(getByA11yState({ expanded: false }, { hidden: true })).toBeTruthy();
+  expect(
```
Or in case you find it useful the current way, just put the `includeHiddenElements` first as the recommended option.

MattAgn replied:
extracting them in another test file sounds good :) (helped me find a bug in the process)

MattAgn commented:
@mdjastrzebski thanks for the review, should be all good now :) 

I've fixed a bug I introduced as well: 
- since we have default values for `defaultIncludeHiddenElements: true`, 
- `configure({ defaultHidden: false })` would get us a config with `{ defaultHidden: false, defaultIncludeHiddenElements: true }``
- in findAll, we would then have  `includeHiddenElements = true` since `getConfig().defaultIncludeHiddenElements`would never be undefined 

That's why I added some code in the configure option to tackle that

mdjastrzebski commented on src/config.ts:
```diff
@@ -14,11 +20,26 @@ export type Config = {
 let config = { ...defaultConfig };
 
 export function configure(options: Partial<Config>) {
+  if (
```
I wonder if we should rather than syncing these two keys have a solution that would set `defaultIncludeHiddenElements` when user passes `defaultHidden`:

```ts
const { defaultHidden, ...restOptions } = options;
const optionsToSet {
  ...restOptions,
  defaultIncludeHiddenElements: restOptions.defaultIncludeHiddenElements ?? defaultHidden ?? config.defaultIncludeHiddenElements
}
```

That way we would also simplify the consuming code. 

Regarding typing this could be handled as:
```ts
interface Config {
  asyncUtilTimeout: number;
  defaultIncludeHiddenElements: boolean;
  // no defaultHidden
}
let config: Config

interface ConfigAliasOptions = {
  defaultHidden?: boolean;
}

function configure(config: Config & ConfigAliasOptions)
```

@MattAgn wdyt?

MattAgn replied:
yes I like that!

mdjastrzebski commented on src/helpers/findAll.ts:
```diff
@@ -4,6 +4,7 @@ import { isHiddenFromAccessibility } from './accessiblity';
 
 interface FindAllOptions {
   hidden?: boolean;
+  includeHiddenElements?: boolean;
```
```suggestion
  includeHiddenElements?: boolean;
  /** RTL-compatible alias to `includeHiddenElements` */
  hidden?: boolean;
```

mdjastrzebski reviewed (approved):
👍🏻
